flareon/src/static_files.rs

```diff
@@ -0,0 +1,385 @@
+//! Static files middleware.
+//!
+//! This middleware serves static files from the `static` directory of the
+//! project.
+
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use bytes::Bytes;
+use futures_core::ready;
+use http::{header, Request};
+use pin_project_lite::pin_project;
+use tower::Service;
+
+use crate::response::{Response, ResponseExt};
+use crate::{AppContext, Body, FlareonApp};
+
+/// Macro to define static files by specifying their paths.
+///
+/// The files are included at compile time using the `include_bytes!` macro.
+/// The paths are relative to the `static` directory of the project (under the
+/// project root, where the `Cargo.toml` file is).
+///
+/// This is mainly useful with the [`FlareonApp::static_files`] trait method.
+///
+/// # Example
+///
+/// ```
+/// use bytes::Bytes;
+/// use flareon::{static_files, FlareonApp};
+///
+/// pub struct ExampleApp;
+///
+/// // Project structure:
+/// // .
+/// // ├── Cargo.toml
+/// // └── static
+/// //     └── admin
+/// //         └── admin.css
+///
+/// impl FlareonApp for ExampleApp {
+///     fn name(&self) -> &str {
+///         "test_app"
+///     }
+///
+///     fn static_files(&self) -> Vec<(String, Bytes)> {
+///         static_files!("admin/admin.css")
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! static_files {
+    ($($path:literal),*) => {
+        vec![$(
+            (
+                $path.to_string(),
+                $crate::__private::Bytes::from_static(
+                    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/static/", $path))
```

Comment:
I don't think those files should be embedded in the binary. First of all, it bloats the binary and we have no idea how many how big files users might want to use. Secondly, it doesn't allow for dynamic file changes (it's rare use case, but I think a valid one) as this happens at compile-time (eg. replacing logo file).

I think we either need to load them dynamically or provide a way to choose between embedding and dynamic loading (eg. two separate macros, or flag in the existing one).

Reply:
Agreed, although it's non-trivial. TODO at https://github.com/orgs/flareon-rs/projects/2/views/7?pane=issue&itemId=84530698